code/__HELPERS/text.dm

```diff
@@ -414,34 +414,159 @@ GLOBAL_LIST_INIT(binary, list("0","1"))
+
+	t = replacetext(t, "!", "$10")
+
+	// Parse hr and small
+
+	if(!limited)
+		t = replacetext(t, "((", "<font size=\"1\">")
+		t = replacetext(t, "))", "</font>")
+		t = replacetext(t, regex("^-{3,}$", "gm"), "<hr>")
+		t = replacetext(t, regex("^\\(-{3,})$", "gm"), "$1")
+
+		// Parse lists
+
+		var/list/tlist = splittext(t, "\n")
+		var/tlistlen = tlist.len
+		var/listlevel = -1
+		var/singlespace = -1 // if 0, double spaces are used before asterisks, if 1, single are
+		for(var/i = 1, i <= tlistlen, i++)
+			var/line = tlist[i]
+			var/count_asterisk = length(replacetext(line, regex("\[^\\*\]+", "g"), ""))
+			if(count_asterisk % 2 == 1 && findtext(line, regex("^\\s*\\*", "g"))) // there is an extra asterisk in the beggining
+
+				var/count_w = length(replacetext(line, regex("^( *)\\*.*$", "g"), "$1")) // whitespace before asterisk
+				line = replacetext(line, regex("^ *(\\*.*)$", "g"), "$1")
+
+				if(singlespace == -1 && count_w == 2)
+					if(listlevel == 0)
+						singlespace = 0
+					else
+						singlespace = 1
+
+				if(singlespace == 0)
+					count_w = count_w % 2 ? round(count_w / 2 + 0.25) : count_w / 2
+
+				line = replacetext(line, regex("\\*", ""), "<li>")
+				while(listlevel < count_w)
+					line = "<ul>" + line
+					listlevel++
+				while(listlevel > count_w)
+					line = "</ul>" + line
+					listlevel--
+
+			else while(listlevel >= 0)
+				line = "</ul>" + line
+				listlevel--
+
+			tlist[i] = line
+		// end for
+
+		t = tlist[1]
+		for(var/i = 2, i <= tlistlen, i++)
+			t += "\n" + tlist[i]
+
+		while(listlevel >= 0)
+			t += "</ul>"
+			listlevel--
+
 	else
-		t = replacetext(t, "\[sign\]", "")
-	t = replacetext(t, "\[field\]", "<span class=\"paper_field\"></span>")
-
-	t = replacetext(t, "\[*\]", "<li>")
-	t = replacetext(t, "\[hr\]", "<HR>")
-	t = replacetext(t, "\[small\]", "<font size = \"1\">")
-	t = replacetext(t, "\[/small\]", "</font>")
-	t = replacetext(t, "\[list\]", "<ul>")
-	t = replacetext(t, "\[/list\]", "</ul>")
+		t = replacetext(t, "((", "")
+		t = replacetext(t, "))", "")
+
+	// Parse headers
+
+	t = replacetext(t, regex("^#(?!#) ?(.+)$", "gm"), "<h2>$1</h2>")
+	t = replacetext(t, regex("^##(?!#) ?(.+)$", "gm"), "<h3>$1</h3>")
+	t = replacetext(t, regex("^###(?!#) ?(.+)$", "gm"), "<h4>$1</h4>")
+	t = replacetext(t, regex("^#### ?(.+)$", "gm"), "<h5>$1</h5>")
+
+	// Parse most rules
+
+	t = replacetext(t, regex("\\*(\[^\\*\]*)\\*", "g"), "<i>$1</i>")
+	t = replacetext(t, regex("_(\[^_\]*)_", "g"), "<i>$1</i>")
+	t = replacetext(t, "<i></i>", "!")
+	t = replacetext(t, "</i><i>", "!")
+	t = replacetext(t, regex("\\!(\[^\\!\]+)\\!", "g"), "<b>$1</b>")
+	t = replacetext(t, regex("\\^(\[^\\^\]+)\\^", "g"), "<font size=\"4\">$1</font>")
+	t = replacetext(t, regex("\\|(\[^\\|\]+)\\|", "g"), "<center>$1</center>")
+	t = replacetext(t, "!", "</i><i>")
+
+	return t
+
+/proc/parsemarkdown_basic_step2(t)
+	if(length(t) <= 0)
+		return
+
+	// Restore the single characters used
+
+	t = replacetext(t, "$10", "!")

```

Comment:
What if someone typed in !0 it would evaluate to $10 when it should be $1

Reply:
Line 439 replaces `!` with `$10`
Now the string is `$100`
Line 525 replaces `$10` with `!`
Now the string is `!0`

But yeah if someone types `\\0` then it will be outputed as `!`. Let me work on this.